Alex/12/AOC2021_12B_v00.py

```diff
@@ -67,88 +67,51 @@ def can_i_visit_this_small_cave(small_cave,path_count):
 			if cave.islower() is True and cave_occurances >= 2:
 				#Another cave has been visited twice. We cannot visit our small cave again.
 				visit_small_cave_bool = False
-		
-		if visit_small_cave_bool == '':
-			#no other small cave has occured twice. We may visit this one again.
-			visit_small_cave_bool = True
-		
-	else:
-		#Small cave has not been visited. You may visit it
-		visit_small_cave_bool = True
-	
-	
-	
 	return visit_small_cave_bool
 
 
-
-
-def produce_new_paths(path,nodesdict,paths,path_count):
+def produce_new_paths(path,nodesdict):
 	#Produce new paths from path
 	
-	new_paths = []
+	new_paths = set()
 	
 	connected_nodes = nodesdict[path[-1]]
 	
 	for node in connected_nodes:
 		if node == 'start':
 			continue
 		
-		if node.islower() is True:
+		if node.islower():
+			path_count = Counter(path)
 			visit_small_cave_bool = can_i_visit_this_small_cave(node,path_count)
 			
 			if visit_small_cave_bool == False:
 				continue
 		
-		new_path = path + [node]
-		
-		if new_path in paths:
-			continue
-		else:
-			new_paths.append(new_path)
+		new_path = *path, node
 		
+		new_paths.add(new_path)
 	
 	return new_paths
 
-def check_all_paths_are_ended(paths):
-	
-	end_nodes = []
-	
-	for path in paths:
-		if path[-1] == 'end':
-			end_nodes.append(True)
-		else:
-			end_nodes.append(False)
-	
-	bool = all(end_nodes)
-	
-	return bool
 
 def find_all_small_town_x1_paths(nodesdict):
-	
-	paths = [['start']]
-	len_paths_old = 0
-	
-	while check_all_paths_are_ended(paths) is False:
-		len_paths_old = len(paths)
-		
+	paths = set()
+	unchecked_paths = set([("start",)])
+
+	while unchecked_paths:
 		new_paths = []
 		
-		for path in paths:
-			if path[len(path)-1] == 'end':
-				new_paths = new_paths + [path]
+		for path in unchecked_paths:
+			if path[-1] == 'end':
+				paths.add(path)
 				continue
 			
 			#Count occurances of caves in path
-			path_count = Counter(path)
-			
-			paths_found = produce_new_paths(path,nodesdict,paths,path_count)
-			new_paths = new_paths + paths_found
-			
-			
-			
+			paths_found = produce_new_paths(path,nodesdict)
+			new_paths.append(paths_found)
 			
-		paths = new_paths
+		unchecked_paths = set().union(*new_paths)

```

Comment:
Nice little hack here to speed up combining sets - combining them all at once (turning a list of sets into a single set) is a fair bit quicker than merging them one by one as you go.

Reply:
In this context, '*' is used to 'unpack' the list - essentially, `func(*[a, b, c])` does `func(a, b, c)`